src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void StartRequest(RequestCompletionSource state)
```

Comment:
My suggestion of buffering may be the wrong one, but at a higher-level the current direction of all of this blocking is going to put us in a really bad place and in my opinion is not feasible for shipping.

If there's really no API provided by libcurl that's entirely callback-based, seems like there a few options:
1) Change to a different implementation, including one based on sockets
2) Build your own event-based system on top of curl_multi_perform.  From the info you provided, it looks very similar to an IOCP; some small number of threads (maybe 1, maybe a handful) block waiting for events queued from a much larger number of sources, and then when one of those has data available, those threads kick off a callback to process that data.